### ISSUE TITLE
How do I train 640x384 images on YOLOv8?(for NPU)

### ISSUE THREAD

**Issue by playdourbest** [ultralytics/ultralytics#16617]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I intend to use the trained model as an inference model for NPU. However, the NPU can only receive 640x384 inputs, so when converting the model to onnx, the input part must be modified to 640x384. So i decided to train with 640x384 data for high mAP. However, I found that rect=false resized the original data from 1280x720 to 640x360 and added padding to learn with 640x640. and rect=True resized 1280x720 of the original data to 640x360 and learned 640x384 by adding padding, finding 32x matching of the shortest. So aren't both methods learned with 640x360? (The data I have is 1280x720.)

How do train from scratch with 640x384? Do I need to modify the code to learn only about 640x384 areas in the v8DetectionLoss class? And do I need to modify the anchor box accordingly? What are the other options?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @playdourbest, thank you for reaching out to Ultralytics 🚀!

We recommend visiting our [Docs](https://docs.ultralytics.com) for guidance on using YOLOv8. You can explore examples for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/), which might answer your questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it effectively.

For your custom training ❓ question, consider providing detailed information such as dataset examples and training logs. Ensure you’re aligning with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our Ultralytics community for more support:
- [Discord](https://ultralytics.com/discord) 🎧 for real-time discussions.
- [Discourse](https://community.ultralytics.com) for in-depth conversations.
- [Subreddit](https://reddit.com/r/ultralytics) for community engagement.

## Upgrade

Ensure you’ve upgraded to the latest version to see if your issue persists:

```bash
pip install -U ultralytics
```

## Environments

YOLO can run in these environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python≥3.8](https://www.python.org/), and [PyTorch≥1.8](https://pytorch.org/) pre-installed):

- **Notebooks**:
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>
  <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
  <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud platforms**: [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) and [Amazon AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: See our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

A green badge means all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are successful. These tests verify the functionality of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across different operating systems.

This is an automated response, but an Ultralytics engineer will assist you soon. Thanks for your patience! 😊

**Comment by Y-T-G** [comment]
That's the intended behavior. YOLO doesn't stretch the image. It preserves the aspect ratio. You can still export it at 640x384.

**Comment by Y-T-G** [comment]
> so when converting the model to onnx, the input part must be modified to 640x384. 

YOLO is fully convolutional network. It doesn't use a fixed input size. You can export your trained model at any input size. However, the best performance would be with the `imgz` you trained with. In this case `640x384`.

**Comment by Y-T-G** [comment]
If your original image is 1280x720, then you should resize to 640x360 to reserve aspect ratio and then add padding to make it 640x640 if you exported with `imgsz=640` or 640x384 if you exported with `imgsz=(384,640)`.

A custom LetterBox implementation can be found here:
https://github.com/ultralytics/ultralytics/blob/f5a491cd45063eea584cccc1e2d4df5575b64057/examples/YOLOv8-OpenCV-int8-tflite-Python/main.py#L17

**Comment by Y-T-G** [comment]
> As you answered
model.train(data='my/data/path',
epochs=epochs,
imgsz=(384, 640),)
Do you mean to make them learn like this?

No, you can't specify width and heigh during training. Only during `model.export()`. You can train with `imgsz=640` and then export with `imgsz=(384,640)`.

**Comment by playdourbest** [comment]
> If your original image is 1280x720, then you should resize to 640x360 to reserve aspect ratio and then add padding to make it 640x640 if you exported with `imgsz=640` or 640x384 if you exported with `imgsz=(384,640)`.
> 
> A custom LetterBox implementation can be found here:
> 
> https://github.com/ultralytics/ultralytics/blob/f5a491cd45063eea584cccc1e2d4df5575b64057/examples/YOLOv8-OpenCV-int8-tflite-Python/main.py#L17

What I know is that YOLOv8 is designed to learn well from square (e.g., 640x640) data.

Does learning with imgsz= (384, 640) using images converted to 640x384 apply YOLO's intended learning method? Is there an operation code to refer to?

**Comment by playdourbest** [comment]
> > As you answered
> > model.train(data='my/data/path',
> > epochs=epochs,
> > imgsz=(384, 640),)
> > Do you mean to make them learn like this?
> 
> No, you can't specify width and heigh during training. Only during `model.export()`. You can train with `imgsz=640` and then export with `imgsz=(384,640)`.

So, is there any specific way to learn 640x384 data?
Changing the aspect ratio to 640x384 instead of 640x360 causes distortion, do you think the approach of re-creating label data and modifying the v8DetectionLoss class and anchor box code is not appropriate?
Is there any other way?

**Comment by Y-T-G** [comment]
Your data has a 16:9 aspect ratio, so it's correctly learning by resizing to 640x360 first and then padding to 384. You can still export at 640x384 imgsz. What's the issue?

If you want to learn without padding, then your data and image should be in a 4:3 aspect ratio, not 16:9. Because 640x384 is 4:3 aspect ratio.

**Comment by playdourbest** [comment]
> Your data has a 16:9 aspect ratio, so it's correctly learning by resizing to 640x360 first and then padding to 384. You can still export at 640x384 imgsz. What's the issue?
> 
> If you want to learn without padding, then your data and image should be in a 4:3 aspect ratio, not 16:9. Because 640x384 is 4:3 aspect ratio.

The model for inference is operated by the NPU, but the output is processed by the CPU. Here, the reshape operation is not available in our NPU, so it works only when the reshape of the HEAD output is removed using the ONNX modifier and the input shape is converted to 640x384.

And I am using a tool that converts the model to binary so that it can work in NPU. If you set import() to imgsz=(384,640) in the export() you told me and put it in the tool after onnx conversion, an error occurs, and if ONNX modifier converts the input to 640x384, it works.

So we approached that learning YOLO with 640x384 would give us a good model, but I think the approach was wrong in the first place because the data we have is 16:9.

**Comment by glenn-jocher** [comment]
To train directly on 640x384 without distortion, consider cropping your images to a 4:3 aspect ratio before training. This way, you can avoid padding and maintain the correct input shape for your NPU.

**Comment by playdourbest** [comment]
> To train directly on 640x384 without distortion, consider cropping your images to a 4:3 aspect ratio before training. This way, you can avoid padding and maintain the correct input shape for your NPU.

So, if i assume that i made 640x384 data with a perfect 4:3 aspect ratio, if i train it with 'rect=True', I will train the image for 640x384?

I understand that the existing YOLO has a 32x matching feature, and if you have a size 360, you can add padding to match it with 384. If I set it to 'rect=True', will this padding feature be omitted?

**Comment by glenn-jocher** [comment]
With `rect=True`, YOLO will resize images to fit within 640x384 while maintaining the aspect ratio, adding padding only if necessary. The 32-pixel alignment is still applied, so ensure your data aligns with this requirement.

**Comment by playdourbest** [comment]
> With `rect=True`, YOLO will resize images to fit within 640x384 while maintaining the aspect ratio, adding padding only if necessary. The 32-pixel alignment is still applied, so ensure your data aligns with this requirement.

Thank you, and I'm glad that some issues have been resolved

**Comment by playdourbest** [comment]
> Your data has a 16:9 aspect ratio, so it's correctly learning by resizing to 640x360 first and then padding to 384. You can still export at 640x384 imgsz. What's the issue?
> 
> If you want to learn without padding, then your data and image should be in a 4:3 aspect ratio, not 16:9. Because 640x384 is 4:3 aspect ratio.

Thank you, and I'm glad that some issues have been resolved

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions or need further assistance, feel free to ask.

**Comment by 13501714030** [comment]
> To train directly on 640x384 without distortion, consider cropping your images to a 4:3 aspect ratio before training. This way, you can avoid padding and maintain the correct input shape for your NPU.

most of my sample photos are 16:9 , i trained directly on 640 x 384 , is it ok?

or if i trained on 640 x 640 and export to 640 x 384, which one is better?

thank you

**Comment by glenn-jocher** [comment]
Training directly on 640x384 is fine if it matches your deployment needs. If your NPU requires 640x384, training at this size may be more efficient.
